examples/05_tool_inference_lock/ex_tool_inference_lock/main.rb

```diff
@@ -0,0 +1,233 @@
+        else
+          # If no face was created we let the user chain new edges to the last
+          # input point.
+          @picked_first_ip.copy!(@mouse_ip)
+        end
+
+        update_ui
+        view.invalidate
+      end
+
+      def onLButtonUp(flags, x, y, view)
+        if @mouse_down.distance([x, y]) > DRAG_THRESHOLD
+          create_edge
+          reset_tool
+        end
+      end
+
+      # Here we have hard coded a special ID for the pencil cursor in SketchUp.
+      # Normally you would use `UI.create_cursor(cursor_path, 0, 0)` instead
+      # with your own custom cursor bitmap:
+      #
+      #   CURSOR_PENCIL = UI.create_cursor(cursor_path, 0, 0)
+      CURSOR_PENCIL = 632
+
+      def onSetCursor
+        # Note that `onSetCursor` is called frequently so you should not do much
+        # work here. At most you switch between different cursor representing
+        # the state of the tool.
+        UI.set_cursor(CURSOR_PENCIL)
+      end
+
+      def onUserText(text, view)
+        distance = text.to_l
+        direction = picked_points[1] - picked_points[0]
+        end_point = picked_points[0].offset(direction, distance)
+        
+        # Can't directly set the position for an InputPoint so we copy it from
+        # a temporary object.
+        # Avoid redefining @mouse_ip as InferenceLockHelper has an internal'
+        # reference to it.
+        @mouse_ip.copy!(Sketchup::InputPoint.new(end_point))
+
+        # If faces are created when creating the edge, reset the tool.
+        # Otherwise keep drawing edges.
+        if create_edge > 0
+          reset_tool
+        else
+          @picked_first_ip.copy!(@mouse_ip)
+        end
+
+        view.invalidate
+      rescue ArgumentError
+        Ui.messagebox('Invalid length')
+      end
+
+      def draw(view)
+        draw_preview(view)
+        
+        view.line_width = 1
+        @mouse_ip.draw(view) if @mouse_ip.display?
+      end
+
+      def getExtents
+        bounds = Geom::BoundingBox.new
+        bounds.add(picked_points)
+        bounds
+      end
+
+      private
+
+      def update_ui
+        # We specifically do not bloat the status text with all the arrow keys
+        # and Shift key, as those are more or less universal in SketchUp.
+        # Adding too much text just makes the status bar hard to read.
+        # However, the instructor should mention these.
+        if picked_first_point?
+          Sketchup.status_text = 'Select end point.'
+        else
+          Sketchup.status_text = 'Select start point.'
+        end
+
+        Sketchup.vcb_label = 'Length'
+      end
+
+      def reset_tool
+        @picked_first_ip.clear
+        
+        # Unlock inference when resetting the tool.
+        # If the user isn't happy with the starting point, they likely aren't
+        # happy with any inference lock either.
+        @inference_lock_helper.unlock
+        
+        update_ui
+      end
+
+      def picked_first_point?
```

Comment:
With the IPs owned by the inference helper this would be a method on the helper. I think perhaps `picked_points` underneath belongs there as well.

This leaves the tool to delegate all inference logic to the helper and then use it's results to create the edge.